### PR TITLE
108: Working with GPUs in OpenVINO notebook

### PR DESCRIPTION
**Pull request by juliomorero**
This notebook provides a high-level overview of working with Intel GPUs in OpenVINO. It shows users how to use Query Device to list system GPUs and check their properties, and it explains some of the key properties. It shows how to compile a model on GPU with performance hints and how to use multiple GPUs using MULTI or CUMULATIVE_THROUGHPUT.

The notebook also shows example commands for benchmark_app that users can run to compare GPU performance in different configurations. It also provides code for a basic end-to-end application that compiles a model on GPU and uses it to run inference.

CC: @helena-intel @ryanloney

### CODE REVIEW

**Comment by ryanloney** [comment]
@juliomorero looks great! I will do a more detailed review with suggested edits. I asked colleagues for review as well.

cc @paularamo @raymondlo84 @adrianboguszewski @eaidova

**Comment by raymondlo84** [comment]
!benchmark_app -m {model_path} -d CPU -hint latency  

we may want to time-bound these (maybe run them for 30 seconds or 15 seconds?).

**Comment by raymondlo84** [comment]
Do we talk about caching to reduce the model compiling time for larger models?

File: notebooks/108-gpu-device/108-gpu-device.ipynb
Comment (eaidova) [inline notebooks/108-gpu-device/108-gpu-device.ipynb:1]: <p>Line #4.    model_path = f"model/public/{model_name}/{precision}/{model_name}.xml"</p><p>to be more consistent, you can reuse downloaded_model_path here as part of model path</p><p> </p><p>model_path = f"{downloaded_model_path}/{precision}/{model_name}.xml"</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (eaidova) [inline notebooks/108-gpu-device/108-gpu-device.ipynb:1]: <p>I think it is good place also for link on AUTO device tutorial too</p><p>https://github.com/openvinotoolkit/openvino_notebooks/tree/main/notebooks/106-auto-device</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

Comment (eaidova) [inline notebooks/108-gpu-device/108-gpu-device.ipynb:1]: <p>I'm not sure that regular user has multiple gpu in setup (e.g. my working machine does not have intel gpu at all :)), can we add some checks before this line to run it conditionally or make this code snippet is part of markdown to prevent unexpected errors? I think it is also good to mention about device naming for multiple gpu in query device section</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

Comment (eaidova) [inline notebooks/108-gpu-device/108-gpu-device.ipynb:1]: <p> please add link on async api tutorial for better explanation of this part</p><p>https://github.com/openvinotoolkit/openvino_notebooks/tree/main/notebooks/115-async-api</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='67'/>

Comment (eaidova) [inline notebooks/108-gpu-device/108-gpu-device.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    def completion_callback(infer_request: InferRequest, results) -&gt; None:
</pre><p>it can be dangerous to store outputs in list here, just because async api does not guarantee consequential execution, you just can draw detections on wrong frame later. Please consider to provide frame id for callback</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='69'/>

**Comment by eaidova** [comment]
> Do we talk about caching to reduce the model compiling time for larger models?

yes, I think model caching is inportant topic for gpu.
Also, not sure, it is already released as part 2022.3 or not, but possible, it will be useful to covert inference precision hints, not only performance hints

**Comment by juliomorero** [comment]
> !benchmark_app -m {model_path} -d CPU -hint latency
> 
> we may want to print the wait time. It would take a while to run and would have no output when the benchmarking is executing.

what do you mean by this? print a message before saying how long it should take? 

I inspected the options for `benchmark_app` and couldn't find one that shows the execution time in real time, the time limit is already shown in step 10 though, for example
```
[Step 10/11] Measuring performance (Start inference asynchronously, 1 inference requests, limits: 60000 ms duration)
[ INFO ] Benchmarking in inference only mode (inputs filling are not included in measurement loop).
[ INFO ] First inference took 32.29 ms
```

File: notebooks/108-gpu-device/108-gpu-device.ipynb
Comment (EdjeElectronics) [reply notebooks/108-gpu-device/108-gpu-device.ipynb:1]: Hi @eaidova , can you please explain how to implement this? The code from this notebook is based off the [classification_sample_async.py](https://github.com/openvinotoolkit/openvino/blob/master/samples/python/classification_sample_async/classification_sample_async.py) sample. The 115-async-api notebook doesn't use AsyncInferRequest (which I think is the preferred way to implement asynchronous inferencing), and it's hard to understand how the code works in that one.

**Comment by EdjeElectronics** [comment]
> yes, I think model caching is inportant topic for gpu. Also, not sure, it is already released as part 2022.3 or not, but possible, it will be useful to covert inference precision hints, not only performance hints

Do you know where we can find documentation on inference precision hints? I'm not sure how it works or what it does. This is the only documentation I can find (from the [Device properties](https://docs.openvino.ai/latest/groupov_runtime_cpp_prop_api.html) page):
image

Thanks very much for the review!

File: notebooks/108-gpu-device/108-gpu-device.ipynb
Comment (EdjeElectronics) [reply notebooks/108-gpu-device/108-gpu-device.ipynb:1]: Nevermind!! Looks like @juliomorero got it figured out 😸

File: notebooks/108-gpu-device/108-gpu-device.ipynb
Comment (ryanloney) [inline notebooks/108-gpu-device/108-gpu-device.ipynb:40]:
```suggestion
    "OpenVINO supports inference on Intel integrated GPUs (which are included with most [Intel® Core™ desktop and mobile processors](https://www.intel.com/content/www/us/en/products/details/processors/core.html)) or on Intel discrete GPU products like the [Intel® Arc™ A-Series Graphics cards](https://www.intel.com/content/www/us/en/products/details/discrete-gpus/arc.html) and [Intel® Data Center GPU Flex Series](https://www.intel.com/content/www/us/en/products/details/discrete-gpus/data-center-gpu/flex-series.html). To get started, first [install OpenVINO](https://docs.openvino.ai/latest/openvino_docs_install_guides_overview.html) on a system equipped with one or more Intel GPUs. Follow the [GPU configuration instructions](https://docs.openvino.ai/latest/openvino_docs_install_guides_configurations_for_intel_gpu.html) to configure OpenVINO to work with your GPU. Then, read on to learn how to accelerate inference with GPUs in OpenVINO!"
```

Comment (ryanloney) [reply notebooks/108-gpu-device/108-gpu-device.ipynb:40]: When using full product names we are required to include (R) and (TM). I updated to also include Data Center Flex GPU (codenamed Arctic Sound) and clarified that iGPU are included in most Core desktop and mobile processors (Xeon processors do not include iGPU).

**Review by ryanloney — CHANGES_REQUESTED** [review]
Very minor edits

File: notebooks/108-gpu-device/108-gpu-device.ipynb
Comment (juliomorero) [reply notebooks/108-gpu-device/108-gpu-device.ipynb:40]: Great, thanks Ryan! 

Now that you mention it (we forgot to add it in the PR description), how should we handle the (R) and (TM) symbols across the notebook? Should we always use them each time Intel/OpenVINO is mentioned or just on product mentions like you did?

File: notebooks/108-gpu-device/108-gpu-device.ipynb
Comment (ngaloppo) [inline notebooks/108-gpu-device/108-gpu-device.ipynb:1]: <p>I believe we should be promoting the MO Ptyhon API from Python code like this...</p><p><a href="https://docs.openvino.ai/latest/openvino_docs_MO_DG_Python_API.html" target="_blank">https://docs.openvino.ai/latest/openvino_docs_MO_DG_Python_API.html</a></p><p> </p><p>At some point there was some work on an OMZ model downloading Python API as well, but that never got merged: <a href="https://github.com/openvinotoolkit/open_model_zoo/pull/1352" target="_blank">https://github.com/openvinotoolkit/open_model_zoo/pull/1352</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/openvinotoolkit/openvino_notebooks/pull/788/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (raymondlo84) [reply notebooks/108-gpu-device/108-gpu-device.ipynb:1]: @eaidova was thinking the same, as we shall update other notebooks so the formatting and style will be similar.

Comment (eaidova) [reply notebooks/108-gpu-device/108-gpu-device.ipynb:1]: yes, I think it is my responsibility as notebooks maintainers to update on new API, I plan to make it after the merge. for now, let's focus on GPU not to make the tutorial overcomplicated with non-device-related stuff :)

**Comment by juliomorero** [comment]
Just so you know, I pushed some commits tackling most of your comments, please let me know what you think about them

On a side note, as you suggested to include model caching stuff, while reading the docs I noticed that the python snippets in the [Even faster](https://docs.openvino.ai/latest/openvino_docs_OV_UG_Model_caching_overview.html#even-faster-use-compile-model-modelpath) section have an inexistent argument `model_path` as it gives the following error:

Screenshot 2023-01-20 at 1 19 48 PM

**Comment by ngaloppo** [comment]
> On a side note, as you suggested to include model caching stuff, while reading the docs I noticed that the python snippets in the [Even faster](https://docs.openvino.ai/latest/openvino_docs_OV_UG_Model_caching_overview.html#even-faster-use-compile-model-modelpath) section have an inexistent argument `model_path` as it gives the following error:

You are using the `openvino.runtime.Core.compile_model` method, which indeed does not have the `model_path` named argument. Instead, use the `openvino.runtime.compile_model` method, which *does* have the `model_path` named argument ([docs](https://docs.openvino.ai/latest/api/ie_python_api/_autosummary/openvino.runtime.compile_model.html)).

**Comment by juliomorero** [comment]
> > On a side note, as you suggested to include model caching stuff, while reading the docs I noticed that the python snippets in the [Even faster](https://docs.openvino.ai/latest/openvino_docs_OV_UG_Model_caching_overview.html#even-faster-use-compile-model-modelpath) section have an inexistent argument `model_path` as it gives the following error:
> 
> You are using the `openvino.runtime.Core.compile_model` method, which indeed does not have the `model_path` named argument. Instead, use the `openvino.runtime.compile_model` method, which _does_ have the `model_path` named argument ([docs](https://docs.openvino.ai/latest/api/ie_python_api/_autosummary/openvino.runtime.compile_model.html)).

The [docs](https://docs.openvino.ai/latest/api/ie_python_api/_autosummary/openvino.runtime.Core.html#openvino.runtime.Core.compile_model) for `openvino.runtime.Core.compile_model` seem to show that the path is supported though, it just uses the same `model` kwarg, right? 

Concerning this notebook, we don't mention the `model_path` stuff so we're ok

**Comment by ngaloppo** [comment]
> The [docs](https://docs.openvino.ai/latest/api/ie_python_api/_autosummary/openvino.runtime.Core.html#openvino.runtime.Core.compile_model) for `openvino.runtime.Core.compile_model` seem to show that the path is supported though, it just uses the same `model` kwarg, right? If that's the case, we just need to replace `model_path` with `model` in those docs
> 
> Concerning this notebook, we don't mention the `model_path` stuff so we're ok

You can just remove the keyword, and it works either way.

**Comment by paularamo** [comment]
@juliomorero Could you please add throughput and the device(s) utilized in the output video?

image

**Comment by paularamo** [comment]
image
Result with:
compiled_model = core.compile_model(model=model, device_name="MULTI:GPU.1,GPU.0,CPU", config={"PERFORMANCE_HINT": "THROUGHPUT"})

**Comment by paularamo** [comment]
image

Results with 
compiled_model = core.compile_model(model=model, 
                                    device_name="AUTO", 
                                    config={"PERFORMANCE_HINT": "THROUGHPUT"})

**Comment by paularamo** [comment]
image

Results using just the dGPU ARC A770 16GB

**Comment by paularamo** [comment]
image

Results using the iGPU

**Comment by paularamo** [comment]
image

Results just using the CPU

**Comment by paularamo** [comment]
Is this an expected behavior of the dGPU? @ngaloppo @ryanloney

**Review by paularamo — CHANGES_REQUESTED** [review]
@juliomorero Could you please add throughput and the utilized device(s) in the output video?

**Comment by ngaloppo** [comment]
> Is this an expected behavior of the dGPU? @ngaloppo @ryanloney @raymondlo84 See the images above. Using the benchmark_app:
> 

@paularamo Are you absolutely sure that `GPU.1` is the dGPU? What kind of platform are you testing this on? Please provide complete `benchmark_app` output.

**Comment by paularamo** [comment]
> Is this an expected behavior of the dGPU? @ngaloppo @ryanloney @raymondlo84 See the images above. Using the benchmark_app:
> 
> image image
> 
> image image
> 
> image image
> 
> I think we new to include -api async in the benchmark command.
> 
> The results using the video example does not show the advantage to use the dGPU because it have the smallest value of FPS. How we can show its total capacity? Could we simulate multiple threads and run multiple inferences?
> 
> Any thought will be great here.

Results with -api async:

image



> > Is this an expected behavior of the dGPU? @ngaloppo @ryanloney @raymondlo84 See the images above. Using the benchmark_app:
> 
> @paularamo Are you absolutely sure that `GPU.1` is the dGPU? What kind of platform are you testing this on? Please provide complete `benchmark_app` output.

Yes @ngaloppo, that is the dGPU:

Here the device info in my machine (Win 11, Gen 12th Core, ARC A770 16GB)

image

image


Here the complete benchmark output:

image
image

**Comment by ngaloppo** [comment]
@paularamo is making some good observations here regarding the throughput of discrete GPUs and how it compares to CPU and GPU. Based on those observations, I suggest considering the following additions: 

1. explore with other models (SSD mobilenet is relatively "cheap") - heavier models will favor the GPU
2. add a section that talks about batching - batching may be required to get better efficiency from discrete GPUs (e.g. when the model is "light"
3. demonstrate the power of batching in the "use case" at the end of the notebook: batch 4 times the video input frame, inference batch-4, and visualize the results in a "4-split mosaic output" hopefully demonstrate that dGPU does this well (better than iGPU/CPU)

**Comment by EdjeElectronics** [comment]
Hi @ngaloppo and @paularamo , I helped @juliomorero with this notebook. We can test out heavier models and see if they provide better results. 

I have a question about batching, though. OpenVINO's [Automatic Batching](https://docs.openvino.ai/latest/openvino_docs_OV_UG_Automatic_Batching.html) feature is supposed to automatically create the best batch sizes for GPU in throughput mode. Since we're using a GPU in throughput mode, shouldn't it already be setting optimal batch sizes in our example?

**Comment by juliomorero** [comment]
> @juliomorero Could you please add throughput (FPS), performance hint, and the utilized device(s) in the output video?

@paularamo you should see them now, I added them in the top left corner. 

I have a question regarding the fps though. In [115-async-api](https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/115-async-api/115-async-api.ipynb) the time is measured from the moment the frame is read until the inference is complete, whereas on [402-pose-estimation](https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/402-pose-estimation-webcam/402-pose-estimation.ipynb) only inference time is measured, similar to what we’re doing. How should we measure the fps? Should we change it?

**Comment by paularamo** [comment]
> > @juliomorero Could you please add throughput (FPS), performance hint, and the utilized device(s) in the output video?
> 
> @paularamo you should see them now, I added them in the top left corner.
> 
> I have a question regarding the fps though. In [115-async-api](https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/115-async-api/115-async-api.ipynb) the time is measured from the moment the frame is read until the inference is complete, whereas on [402-pose-estimation](https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/402-pose-estimation-webcam/402-pose-estimation.ipynb) only inference time is measured, similar to what we’re doing. How should we measure the fps? Should we change it?

@juliomorero Thanks for adding that

image

I think we don't need to change how we are measuring the FPS. Inference time is OK.

**Review by paularamo — APPROVED** [review]

**Comment by paularamo** [comment]
@ryanloney this notebook is ready for your review.

**Comment by ryanloney** [comment]
I don't have any more edits. However, I'm unable to get the output video. It does not save and when I run the final cell I get an error. When commenting out the `Video("output/output.mp4")` there is still a failure to save the .mp4 but no error message. I'm running on Fedora Linux, so maybe there is some issue using this OS.. 

Screenshot from 2023-02-03 17-08-07

**Review by ryanloney — APPROVED** [review]

File: notebooks/108-gpu-device/108-gpu-device.ipynb
Comment (OpenVINO-dev-contest) [inline notebooks/108-gpu-device/108-gpu-device.ipynb:890]: Hi actually the ```time per frame``` here means the throughput, how about latency ? shall we also record the average inference time for each frame ?

Comment (eaidova) [reply notebooks/108-gpu-device/108-gpu-device.ipynb:890]: Let's do it as follow up

File: notebooks/108-gpu-device/108-gpu-device.ipynb
Comment (eaidova) [inline notebooks/108-gpu-device/108-gpu-device.ipynb:786]: probably we can consider to  have some label map for coco dataset in notebooks/data for avoid coping these from tutorial to tutorial

File: notebooks/108-gpu-device/108-gpu-device.ipynb
Comment (eaidova) [inline notebooks/108-gpu-device/108-gpu-device.ipynb:833]:
```suggestion
    "    results[frame_id] = predictions[:10]  # Grab first 10 predictions for this frame\n",
```

File: notebooks/108-gpu-device/108-gpu-device.ipynb
Comment (eaidova) [inline notebooks/108-gpu-device/108-gpu-device.ipynb:882]:
```suggestion
    "infer_queue.wait_all()  # Wait until all inference requests in the AsyncInferQueue are completed\n",
```

**Review by eaidova — APPROVED** [review]

**Pull request merged**
